### DISCUSSION TITLE
Execute only on certain branch names during PR

### DISCUSSION THREAD

**Discussion by ConstantinoSchillebeeckxSimpleRose** [community/community#26531] [Actions]
<p>I’d like for an action to run on any PR that’s opened and that has a head name like feature/**, fix/** or hotfix/**</p>
<pre><code>name: something
on:
  pull_request:
    types:
    - opened
    - reopened
    - synchronize
    branches:
    - develop
jobs:
  something:
    name: Unit tests
    runs-on: ubuntu-18.04
    if: contains(github.head_ref, 'feature/**') || contains(github.head_ref, 'fix/**') || contains(github.head_ref, 'hotfix/**')
    steps:
...
</code></pre>
<p>However this doesn’t work, the action just sits in “Starting your workflow runs…”</p>

**Comment by jhenstridge** [comment]
<p>There’s nothing <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#contains" rel="nofollow noopener">in the documentation</a> to suggest that contains() performs regular expression or pattern matching, so that probably why this isn’t working.</p>
<p>Given that all the patterns you’ve listed are prefix matches, have you considered using <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#startswith" rel="nofollow noopener">startsWith()</a> instead?</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/constantinoschillebe">@constantinoschillebe</a> ,</p>
<p>I add some comments to your code, please kindly check:</p>
<pre><code>name: something
on:
  pull_request:
    types:
    - opened
    - reopened
    - synchronize
    branches:
    - develop #the pull requst will be triggered only when target branch is 'develop'.
jobs:
  something:
    name: Unit tests
    runs-on: ubuntu-18.04
    if: contains(github.head_ref, 'feature') || contains(github.head_ref, 'fix') || contains(github.head_ref, 'hotfix') #remove **, it doesn't accept pattern, or you can use startswith() instead
    steps:
    ...
</code></pre>
<p>The workflow will be triggered when PR is created from ‘feature**,fix**, hotfix**’ branch to ‘develop’ branch.</p>
<p>However i doubt whether you want to check the source branch, if you’d like to limit the workflow to ‘target’ branch,  on ‘<a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestbranchestags" rel="nofollow noopener">pull_request.branches</a>’ is enough.</p>

**Comment by ConstantinoSchillebeeckxSimpleRose** [comment]
<p>Thanks for the suggestions <a class="mention" href="/weide-zhou">@weide-zhou</a>; I’ve updated my code</p>
<p>My use case is as follows: we use actions to run all our unit tests including functional tests, we use gitflow for branch naming and therefore I do not want to run unit tests on branches like “docs/**” or “chore/**”. We use these actions as required status checks (base_ref develop) before merging on all head_ref. Therefore, even if skipping the action, I need it to return some sort of success status. I’m doing that now like this:</p>
<pre><code>name: something
on:
  pull_request:
    types:
    - opened
    - reopened
    - synchronize
    branches:
    - develop #the pull requst will be triggered only when target branch is 'develop'.
jobs:
  something:
    name: Unit tests
    runs-on: ubuntu-18.04
    steps:
- name: step1
if: "! contains(toJSON(github.head\_ref), 'chore/') &amp;&amp; ! contains(toJSON(github.head\_ref), 'docs/')" ...
- name: step2
 if: "! contains(toJSON(github.head\_ref), 'chore/') &amp;&amp; ! contains(toJSON(github.head\_ref), 'docs/')"
...
- name: step3
 if: "! contains(toJSON(github.head\_ref), 'chore/') &amp;&amp; ! contains(toJSON(github.head\_ref), 'docs/')"
...
</code></pre>
<p>I tried using </p>
<pre><code>if:success()
</code></pre>
<p>instead of copy&amp;pasting that if line on each step, but that doesn’t work either.</p>
<p>Ideally I’d like to do an early success termination on the condition below</p>
<pre><code>if: contains(toJSON(github.head_ref), 'chore/') &amp;&amp; contains(toJSON(github.head_ref), 'docs/')
</code></pre>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/constantinoschillebe">@constantinoschillebe</a> ,</p>
<p>Glad to support! Since you’d like to return success even if skipping the action, as your code shown, probably you have to set ‘if’ expression on step level instead of job level to achieve this.</p>
<p>BTW, if you’d like to skip the whole workflow for PR source branch ‘docs/**’ or ‘chore/**’, you can simply remove the yaml file in these branches.</p>

**Comment by ConstantinoSchillebeeckxSimpleRose** [comment]
<blockquote>
<p>Since you’d like to return success even if skipping the action, as your code shown, probably you have to set ‘if’ expression on step level instead of job level to achieve this.</p>
</blockquote>
<p>Unfortunately that doesn’t actually work; if the whole action is skipped, success is never returned to the status.</p>
<blockquote>
<p>BTW, if you’d like to skip the whole workflow for PR source branch ‘docs/**’ or ‘chore/**’, you can simply remove the yaml file in these branches.</p>
</blockquote>
<p>This also doesn’t work as the PR will get merged into develop.</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/constantinoschillebe">@constantinoschillebe</a> ,</p>
<p>I create a repository for checking:</p>
<ol>
<li>
<p>Pull request from ‘docs/B2’ to ‘develop’ branch: <a href="https://github.com/weide-zhou/ticket22/pull/1" rel="nofollow noopener">https://github.com/weide-zhou/ticket22/pull/1</a></p>
</li>
<li>
<p>Pull request from ‘feature/B1’ to ‘develop’ branch: <a href="https://github.com/weide-zhou/ticket22/pull/2" rel="nofollow noopener">https://github.com/weide-zhou/ticket22/pull/2</a></p>
</li>
<li>
<p>Pull request form ‘succeedBranch’ to develop, no step is skiped: <a href="https://github.com/weide-zhou/ticket22/pull/3" rel="nofollow noopener">https://github.com/weide-zhou/ticket22/pull/3</a></p>
</li>
</ol>
<p>Status check is passed and steps are skiped for source branch docs/** and feature/**.</p>
<p>Please let me know if my understanding is wrong.</p>

**Reply by MasoudGhodrati** [reply]
can't open those repos

**Reply by jgu7man** [reply]
x2
